fix juniper wildcard handling

arifogel opened the pull request:
Fix improper pruning of strings that look like wildcards but did not get generated via `apply-groups`.

Implementation:
- Create new virtual token called `WILDCARD_ARTIFACT` that is used instead of `WILDCARD` when generating lines.
- Prune only lines with `WILDCARD_ARTIFACT` instead of `WILDCARD` as before
- Update parser rules to allow `WILDCARD` or `WILDCARD_ARTIFACT` via new rule `wildcard` so parsing during group line generation phase still works

dhalperi commented:
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/ConfigurationMatchersImpl.java, line 161 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwkBco49-RoL-BwiyH:-LDwkBco49-RoL-BwiyI:bv5gpjl) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/ConfigurationMatchersImpl.java#L161)):*
> ```
> RouteFilterLists:", "routeFilterLists"
> ```

what's the style recommendation here? are they supposed to be identical?  (`R` vs `r`?)

---

*[projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_routing_instances.g4, line 229 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwkexG5KKVI0WNLO-r:-LDwkexG5KKVI0WNLO-s:bphu3p) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_routing_instances.g4#L229)):*
> ```
> VARIABLE
> ```

vs `variable`? Remind me how you choose?

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1333 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwlUBbEAQ3cuCt2ZTN:-LDwlUBcDX1RV0dVuZJH:b-bq1m0b) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1333)):*
> ```Java
>     Configuration c = b.loadConfigurations().get(hostname);
> 
>     System.out.println(
> ```

no println

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1554)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 10 of 24 files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/ConfigurationMatchersImpl.java, line 161 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwkBco49-RoL-BwiyH:-LDwxc0s5MMae4Nk7dHp:b-886gsx) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/ConfigurationMatchersImpl.java#L161)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> RouteFilterLists:", "routeFilterLists"
> ```

what's the style recommendation here? are they supposed to be identical?  (`R` vs `r`?)
</blockquote></details>

Fixed

---

*[projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_routing_instances.g4, line 229 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwkexG5KKVI0WNLO-r:-LDwxhDpBV8ePI3lq1XE:bpmpfqa) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_routing_instances.g4#L229)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> VARIABLE
> ```

vs `variable`? Remind me how you choose?
</blockquote></details>

Out of scope, but I generally avoid referring to `VARIABLE` ever, because it improperly prevents names that are built-in keywords. This should be changed to `variable`.

---

*[projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java, line 1333 at r1](https://reviewable.io/reviews/batfish/batfish/1554#-LDwlUBbEAQ3cuCt2ZTN:-LDwxuuw3xylzlt3dg8-:ba3dfob) ([raw file](https://github.com/batfish/batfish/blob/3641c603edff35de3f2577ffdeb1fa55f25991cc/projects/batfish/src/test/java/org/batfish/grammar/flatjuniper/FlatJuniperGrammarTest.java#L1333)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

no println
</blockquote></details>

Removed

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1554)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1554)*
<!-- Sent from Reviewable.io -->
